agent.go

```diff
@@ -315,7 +290,160 @@ func (a *Agent) watchHealthChecks(updateCh chan api.HealthChecks) {
 			continue
 		}
 
+		ourChecks := make(api.HealthChecks, 0)
+		for _, c := range checks {
+			if ourNodes[c.Node] {
+				ourChecks = append(ourChecks, c)
+			}
+		}
+
 		opts.WaitIndex = meta.LastIndex
-		updateCh <- checks
+		updateCh <- ourChecks
+	}
+}
+
+// computeWatchedNodes watches both the list of registered ESM instances and the list of
+// external nodes registered in Consul and decides which nodes the local agent should be
+// in charge of in a deterministic way.
+func (a *Agent) computeWatchedNodes(serviceID string, healthNodeCh chan map[string]bool, pingNodeCh chan []*api.Node) {
+	nodeCh := make(chan []*api.Node)
+	instanceCh := make(chan []*api.ServiceEntry)
+
+	go a.watchExternalNodes(nodeCh)
+	go a.watchServiceInstances(instanceCh)
+
+	externalNodes := <-nodeCh
+	healthyInstances := <-instanceCh
+
+	firstRun := make(chan struct{}, 1)
+	firstRun <- struct{}{}
+	prevCount := -1
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case externalNodes = <-nodeCh:
+		case healthyInstances = <-instanceCh:
+		case <-firstRun:
+			// Skip the wait on the first run.
+		}
+
+		// Get our "index" in the list of sorted service instances
+		ourIndex := -1
+		for i, instance := range healthyInstances {
+			if instance.Service.ID == serviceID {
+				ourIndex = i
+				break
+			}
+		}
+		if ourIndex == -1 {
+			a.logger.Printf("[WARN] Didn't register service with Consul yet, retrying...")
```

Comment:
This could also happen if the current node is put into maint mode or legitimately failing a health check. Maybe we should make this message a little more generic for those cases?